### ISSUE TITLE
Wrong return type inferred for write!() and writeln!()

### ISSUE THREAD

**Issue by kumbayo** [intellij-rust/intellij-rust#2229]
Currently both `write!()` and `writeln!()` macros are inferred to return `()`.
In most cases the macros should actually return a different type.

The implementation of the `write!()` and `writeln!()` macros calls write_fmt() on their first argument and return the result of that call.

When called on a expression implementing the `io::Write` trait it should return `io::Result<()>`
When called on a expression implementing the `fmt::Write` trait it should return `fmt::Result<()>`

This was reported at https://users.rust-lang.org/t/what-is-the-expect-function-for-writeln-macro/15200/

**Comment by kumbayo** [comment]
@vlad20012 Not sure if this is expected with the current state of macro expansion or not.
The expansion currently produces wrong results when writing `&mut s` instead of `s` for code like this (Thats the way it is written in the documentation of `write!()`).

```rust
use std::fmt::Write;
fn main() {
    let mut s = String::new();
    let a = write!(&mut s, "text");
    a;
  //^ IntelliJ detects &mut Result<(), Error> instead of Result<(), Error>
}
```
The expansion currently works on a textual basis so the `&mut` is applied to the whole expression
instead of parsing it as `(&mut s).write_all(...)`

**Comment by vlad20012** [comment]
@kumbayo Ok, I think we can simply fix it by wrapping expressions with parentheses before substituting to a macro
